### DISCUSSION TITLE
Deprecation release

### DISCUSSION THREAD

**Discussion by ipmb** [pydantic/pydantic#4228] [Pydantic V2]
Hi @samuelcolvin I read through [your detailed plan for v2](https://pydantic-docs.helpmanual.io/blog/pydantic-v2/) and since you solicited feedback, here's mine 😄

It might be nice to have the final `1.x` release raise deprecation warnings around support that will be removed in v2. This would allow downstream projects to see where their code needs to change before the upgrade.

I realize it's probably a lot of work, but certainly a nice-to-have for end-users.

Thanks for pydantic! It's a great project and I'm looking forward to a faster v2!

**Comment by samuelcolvin** [comment]
I think that's a good idea, though as you say, quite a lot of work. It won't be v1.10 which will be a standard release.

But maybe we could do a v1.11 release which is just deprecation warnings.

Lots of work, but maybe less work than adding all the deprecation warnings to pydantic V2. @PrettyWood any thoughts?

**Reply by Bobronium** [reply]
I didn't research this much, but I think for smooth transition, It's a must to release version that will be mostly compatible with current v1.x and will emit warnings on access of deprecated things.

We can (ab)use [`__getattr__`](https://peps.python.org/pep-0562/) on modules to track when something deprecated is imported.

And use some fancy decorators along with custom generic types.
Decorators should emit warnings in runtime, while generic types should trigger errors in mypy and editors:
<img width="400" alt="image" src="https://user-images.githubusercontent.com/36469655/212180027-70567cee-c0d4-4014-9cca-194d802b9de3.png">
<img width="509" alt="image" src="https://user-images.githubusercontent.com/36469655/212180030-9cdbdb40-f117-4366-918c-4f61646fb23e.png">

But still, lots of work to do indeed.

One way to approach this is to find/write some handy decorators which can be configured declaratively and will emit warnings in runtime, and then if someone will be willing to do it, write custom generic types and generate `.pyi` stubs, so they will be picked up by typecheckers and IDE's.

**Comment by jaraco** [comment]
Looks like a deprecation release was never made. This meant that libraries like [inflect](/jaraco/inflect) were suddenly broken without warning for very minor uses of pydantic.

I'd heard that a new release of Pydantic was coming but was unconcerned as I hadn't seen any deprecation warnings pointing to changes that would be necessary, so I'd hoped/assumed the effect was a no-op for inflect.
